File: core/src/ten_utils/schema/bindings/rust/schema.c
//
// Copyright © 2025 Agora
// This file is part of TEN Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to the "LICENSE" file in the root directory for more information.
//
#include "include_internal/ten_utils/schema/bindings/rust/schema.h"

#include "include_internal/ten_utils/schema/schema.h"
#include "ten_utils/lib/alloc.h"
#include "ten_utils/lib/error.h"
#include "ten_utils/macro/check.h"

ten_schema_t *ten_schema_create_from_json_str_proxy(const char *json_string,
                                                    const char **err_msg) {
  TEN_ASSERT(json_string, "Invalid argument.");
  return ten_schema_create_from_json_str(json_string, err_msg);
}

void ten_schema_destroy_proxy(const ten_schema_t *self) {
  ten_schema_t *self_ = (ten_schema_t *)self;
  TEN_ASSERT(self, "Invalid argument.");
  TEN_ASSERT(ten_schema_check_integrity(self_), "Invalid argument.");

  ten_schema_destroy(self_);
}

bool ten_schema_adjust_and_validate_json_str_proxy(const ten_schema_t *self,
                                                   const char *json_string,
                                                   const char **err_msg) {
  ten_schema_t *self_ = (ten_schema_t *)self;
  TEN_ASSERT(self, "Invalid argument.");
  TEN_ASSERT(ten_schema_check_integrity(self_), "Invalid argument.");
  TEN_ASSERT(json_string, "Invalid argument.");

  return ten_schema_adjust_and_validate_json_str(self_, json_string, err_msg);
}

bool ten_schema_is_compatible_proxy(const ten_schema_t *self,
                                    const ten_schema_t *target,
                                    const char **err_msg) {
  ten_schema_t *self_ = (ten_schema_t *)self;
  TEN_ASSERT(self, "Invalid argument.");
  TEN_ASSERT(ten_schema_check_integrity(self_), "Invalid argument.");

  ten_schema_t *target_ = (ten_schema_t *)target;
  TEN_ASSERT(target && ten_schema_check_integrity(target_),
             "Invalid argument.");

  ten_error_t err;
  TEN_ERROR_INIT(err);

  bool result = ten_schema_is_compatible(self_, target_, &err);
  if (!result) {
    *err_msg = ten_strdup(ten_error_message(&err));
  }

  ten_error_deinit(&err);
  return result;
}

File: core/src/ten_manager/src/designer/extensions/schema.rs
//
// Copyright © 2025 Agora
// This file is part of TEN Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to the "LICENSE" file in the root directory for more information.
//
use std::sync::{Arc, RwLock};

use actix_web::{web, HttpResponse, Responder};
use anyhow::{anyhow, Result};
use serde::{Deserialize, Serialize};

use ten_rust::pkg_info::manifest::api::ManifestApi;

use crate::designer::{
    response::{ApiResponse, ErrorResponse, Status},
    DesignerState,
};

#[derive(Deserialize, Serialize, Debug)]
pub struct GetExtensionSchemaRequestPayload {
    pub app_base_dir: String,
    pub addon_name: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct GetExtensionSchemaResponseData {
    pub schema: Option<ManifestApi>,
}

pub async fn get_extension_schema_endpoint(
    request_payload: web::Json<GetExtensionSchemaRequestPayload>,
    state: web::Data<Arc<RwLock<DesignerState>>>,
) -> Result<impl Responder, actix_web::Error> {
    let state_read = state.read().map_err(|e| {
        actix_web::error::ErrorInternalServerError(format!(
            "Failed to acquire read lock: {}",
            e
        ))
    })?;

    let pkgs_cache = state_read.pkgs_cache.read().await;

    // Check if app exists.
    let pkgs_info_in_app = match pkgs_cache.get(&request_payload.app_base_dir) {
        Some(info) => info,
        None => {
            let error_response = ErrorResponse::from_error(
                &anyhow!("App not found"),
                "App not found",
            );
            return Ok(HttpResponse::NotFound().json(error_response));
        }
    };

    // Check if extension info exists.
    let extension_pkgs_info = match &pkgs_info_in_app.extension_pkgs_info {
        Some(info) => info,
        None => {
            let error_response = ErrorResponse::from_error(
                &anyhow!("Extension not found"),
                "Extension not found",
            );
            return Ok(HttpResponse::NotFound().json(error_response));
        }
    };

    // Find specific extension.
    let extension_pkg_info = extension_pkgs_info.iter().find(|pkg_info| {
        pkg_info.manifest.type_and_name.name == request_payload.addon_name
    });

    // Check if extension exists.
    let extension_pkg_info = match extension_pkg_info {
        Some(info) => info,
        None => {
            let error_response = ErrorResponse::from_error(
                &anyhow!("Extension not found"),
                "Extension not found",
            );
            return Ok(HttpResponse::NotFound().json(error_response));
        }
    };

    // Success case.
    let response = ApiResponse {
        status: Status::Ok,
        data: GetExtensionSchemaResponseData {
            schema: extension_pkg_info.manifest.api.clone(),
        },
        meta: None,
    };

    Ok(HttpResponse::Ok().json(response))
}
